### ISSUE TITLE
improving distributed logging story

### ISSUE THREAD

**Issue by yaroslavvb** [wandb/wandb#452]
I played around with a few different ways of distributed logging in wandb and I'm learning towards recommending to log from just 1 worker for the PyTorch/AWS [tutorial](https://docs.google.com/document/d/1LuCACKdXTbptHzXb81sGA1CevRW44wQ5znxXylj8S5A/edit#heading=h.d1y77q80usbq)

IE
```
if rank=0:
  wandb.init(...)
```
then replace wandb.log with "wandb_log" where
```
def wandb_log(*args_, **kwargs_):
    if rank=0:
        wandb.log(*args_, **kwargs_)
```
Reasons:
1. 95% of the time, you don't care about logs from other workers, chief worker has all info you need. Having a mix of single runs and grouped runs complicates runs interface. People using TensorBoard only log from chief worker.
2. Using grouping doesn't really work -- if you have 8-worker trainer and run it twice, wandb will show it as 16-worker group. There's a complicated solution in https://github.com/wandb/client/issues/335#issuecomment-493284910 
3. 5% of the time, one worker crashes or is slow, so you want to look at individual workers. In that case it makes sense to log from all workers and do grouping. To avoid merging of groups, use unique group name, ie group_name='debug_attempt_523'

What wandb can do:
- replacing wandb.log with wandb_log is extra work on the user. Currently it is necessary because otherwise wandb.log crashes. One idea: `wandb.init(no_op=True)`, makes all subsequent `wandb.log` return immediately instead of crashing

**Comment by vanpelt** [comment]
Hey @yaroslavvb I could imagine a `disable` keyword arg to `wandb.init`.  Until then you can do the following to achieve what you want:

```python
import os
import wandb

if rank!=0:
    os.environ['WANDB_MODE'] = 'dryrun'

wandb.init(...)
```

**Comment by yaroslavvb** [comment]
Some more thoughts.

- Having 1 run entry per process and using grouping is the wrong abstraction for distributed training. Even though there are several processes, there's still one set of important metrics, like accuracy. Going from 1 GPU to 2 GPUs should feel like a single training run being launched on a bigger computer rather than two separate runs. The latter is what's happening when there's one process per GPU, which is what Horovod or PyTorch DDP do.

The common structure is that there's client process which launches several worker processes, and the first worker process (chief process) is in charge of aggregating statistics and logging them.

# Current solution
Mark one worker as special (chief) and do all the logging from this process. Because the user interacts with client process, but logging is done from chief worker process, some logging information must be propagated from client to chief worker.

![wandb-distributed@2x](https://user-images.githubusercontent.com/23068/62063726-4fbba900-b1e0-11e9-81ba-08bd4675f899.png)

- to capture source information in wandb, client uploads `.git` folder to chief worker. This takes couple of seconds using [rsync](https://github.com/cybertronai/imagenet18_/blob/8f5e0f7fde95167f5c011e2451433b007ff5e4f1/train.py#L169)
- to capture config information from client, client pickles config dict and uploads it. The chief worker unpickles it and [updates](https://github.com/cybertronai/aws-network-benchmarks/blob/f10e1bd9cf2c4dc48ba39a0b5ddce3dd29b4304a/nccl_bench.py#L189) `wandb.config`

# Downsides
- stdout streams and performance metrics from non-chief worker processes are occasionally useful and are not captured
- client stdout logs are occasionally useful but not captured
- client -> chief worker upload complicates user code

# Discussion
- The client process is the only one that can tie all processes in the training run together. Therefore client must generate unique id and propagate it to all processes.
- Want to have access to client + non-chief worker logs, but don't want to clutter run list with separate run entries for them
- Some `wandb.config` entries are filled in by client, but others by chief worker, want them shown on common interface.

# Possible solution
One could tie all the processes together using `group_id` and have one process initialize with `chief=True`.  When one process in the group is initialized with `chief=True`, wandb creates a special "merged" run and hides the rest. The `chief` naming convention is taken from TensorFlow which uses this keyword to mark a special worker in charge of logging/checkpointing.

The merged run: 
- for `wandb.log`, it is a copy of logs  from the chief worker process
- for `wandb.config`, entries are merged from all processes, but `chief` worker values take precedence over the rest when in conflict.
- for `.git` version tracking, use `chief` version takes precedence when processes have conflicting `.git` (should not happen, most likely a bug in user's job setup logic)
- merged run provides links to individual "raw" runs
- Optional: allow workers to skip `project` setting, this can be determined from project setting of client process and `group_id` 

```
client:
id = wandb.util.generate_id()
wandb.init(group_id=id, job_type='client', project='distributed')
wandb.config['val']=val
for i in range(num_procs):
  os.system(f'python worker.py --id={id} --rank={i}')

worker:
wandb.init(group_id=args.id, chief=`args.rank==0`)
wandb.config['someval']=someval
```

**Comment by vanpelt** [comment]
This is a great writeup!  We'll discuss internally.

**Comment by tshrjn** [comment]
Has there been any update on this issue?

**Comment by vanpelt** [comment]
@tshrjn This likely won't be addressed this quarter.  What's the most important part of this discussion for your workflows?

**Comment by tshrjn** [comment]
Thanks for your quick reply, for now, I'm just doing:

```
if rank=0:
  wandb.init(...)
```

**Comment by yaroslavvb** [comment]
I eventually settled on following logic.

1. Use a global "globals.py" module to keep track of global step and global settings. This simplifies things when training/initialization logic is split across several modules.
2. Use wandb through EventWriter writer interface/monkeypatching, otherwise `wandb.log` crashes in workers that didn't call init.

```
globals.py:
token_count: int = 0   
event_writer: Optional[SummaryWriter] = None
args = None   #  global arg values


log.py:
import globals as g

def logging_setup():
    if util.get_global_rank() == 0:
        wandb.init(project="Transformer-XL source code", name=g.args.run_name, sync_tensorboard=True)
        g.event_writer = SummaryWriter(g.args.logdir)
    else:
        g.event_writer = util.NoOp() 

    else:
        g.event_writer.add_scalar(tag, val, g.token_count)

def log_tb(tag, val):
        g.event_writer.add_scalar(tag, val, g.token_count)

util.py:
class NoOp: # no_op method/object that accept every signature
    def __getattr__(self, *_args):
        def no_op(*_args, **_kwargs): pass

        return no_op
```

The downside is that a small fraction of the time you need to troubleshoot non-chief worker. To cover this case, I run all workers in their own tmux sessions, and look at the crash logs by attaching to the appropriate tmux session

**Comment by Hzzone** [comment]
After a long time to try, I found a fast way to combine ddp with wandb, that is re-write `torch.distributed.launch` like below:

```python
import sys
import argparse
import torch
import os
import subprocess
import wandb

#####
from lib.util import get_free_port

parser = argparse.ArgumentParser()
parser.add_argument('--device-ids', help='devices id to be used', type=str)
parser.add_argument("--id", help='restore id', type=str)
# positional
parser.add_argument("training_script", type=str,
                    help="The full path to the single GPU training "
                         "program/script to be launched in parallel, "
                         "followed by all the arguments for the "
                         "training script")

# rest from the training program
parser.add_argument('training_script_args', nargs=argparse.REMAINDER)

if __name__ == '__main__':
    args = parser.parse_args()

    # set PyTorch distributed related environmental variables
    world_size = len(args.device_ids.split(','))
    current_env = os.environ.copy()
    current_env["MASTER_ADDR"] = '127.0.0.1'
    current_env["MASTER_PORT"] = str(get_free_port())
    current_env["WORLD_SIZE"] = str(world_size)

    processes = []
    for local_rank in range(world_size):
        # each process's rank
        current_env["RANK"] = str(local_rank)
        current_env["LOCAL_RANK"] = str(local_rank)

        # spawn the processes
        cmd = [sys.executable, "-u", args.training_script]

        cmd.append("--local_rank={}".format(local_rank))

        # add wandb id
        cmd.append("--id={}".format(args.id if args.id is not None else wandb.util.generate_id()))
        # add device ids
        cmd.append("--device-ids={}".format(args.device_ids))
        # add other args
        cmd.extend(args.training_script_args)

        process = subprocess.Popen(cmd, env=current_env)
        processes.append(process)

    for process in processes:
        process.wait()
        if process.returncode != 0:
            raise subprocess.CalledProcessError(returncode=process.returncode,
                                                cmd=cmd)
```

The id of different processes is supposed to same, and you could modify it for your own need. Note that we should add a `--id` and `--device-ids` to your script like `--local-rank` in `torch.distributed.launch`.

**Comment by aced125** [comment]
I'm experiencing this issue! Of wandb spawning multiple runs when doing ddp training.

Will do the rank zero thing to avoid for now.

**Comment by kaixin96** [comment]
In some cases, collecting logs from each worker is pretty useful. For example, when training RL agents, wandb allows collecting metrics from each actor and learner. So users do not need to write code for retrieving metrics from remotes to the chief worker. Also, actor-specific metrics can stay on remotes, avoiding communication cost.

**Comment by raubitsj** [comment]
@kaixin96 We just released a new library that *will* soon be much easier to support parallel runs (distributed training) much better

We will update this bug when the parallel runs are implemented,tested, and documented.

**Comment by joshlk** [comment]
@raubitsj has does wanadb now support support distributed training? Can you point me in the right direction as I can't find any documentation.

**Comment by kwanUm** [comment]
Following as well

**Comment by kanybekasanbekov** [comment]
Can't wait when WANDB will natively support distributed training

**Comment by vanpelt** [comment]
Hey @kwanUm, @joshlk, @kanybekasanbekov apologies for the delay in response here.  wandb does support distributed training natively.  This thread is about more complex distributed cases.  When training in a distributed manner, you will likely want to use [grouping](https://docs.wandb.ai/library/grouping) to view many different "runs" as a single experiment in the wandb UI.  

If you can describe your specific use cases or limitations of the grouping approach you're looking to solve for we can provide guidance on what functionality is in development.

**Comment by kptkin** [comment]
We have a working solution with the new service feature. Closing this issue.
